src/esp32/hcsr04.rs

```diff
@@ -0,0 +1,289 @@
+            Err(prior) => {
+                // Final edge: try set the lockout to debounce
+                // (i.e. we observe `timestamp` to be still set to
+                // `prior`), and if so, return a reply.
+                if let Ok(_) =
+                    arg.timestamp
+                        .compare_exchange(prior, -1, Ordering::AcqRel, Ordering::Acquire)
+                {
+                    // Notify the waiter if we can convert the
+                    // computed duration into a non-zero u32. If we
+                    // don't notify, the waiter will time out and
+                    // return an error, and the state machine will be
+                    // reset on the next `get_readings` call.
+                    if let Ok(delta) = u32::try_from(when - prior) {
+                        if let Some(nz) = NonZeroU32::new(delta) {
+                            arg.notifier.notify_and_yield(nz);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl Drop for HCSR04Sensor {
+    fn drop(&mut self) {
+        let pin = self.echo_interrupt_pin.pin();
+        if let Err(error) = unsafe { esp!(gpio_isr_handler_remove(pin)) } {
+            log::warn!(
+                "HCSR04Sensor: failed to remove interrupt handler for pin {}: {}",
+                pin,
+                error
+            )
+        }
+    }
+}
+
+impl Sensor for HCSR04Sensor {}
+
+impl Readings for HCSR04Sensor {
+    fn get_generic_readings(&mut self) -> anyhow::Result<GenericReadingsResult> {
+        Ok(self
+            .get_readings()?
+            .into_iter()
+            .map(|v| (v.0, SensorResult::<f64> { value: v.1 }.into()))
+            .collect())
+    }
+}
+
+impl SensorT<f64> for HCSR04Sensor {
+    fn get_readings(&mut self) -> anyhow::Result<TypedReadingsResult<f64>> {
+        // If the echo pin is already high for some reason, the state machine
+        // won't work correctly.
+        if self.echo_interrupt_pin.is_high() {
+            anyhow::bail!("HCSR04Sensor: echo pin is high before trigger sent")
+        }
+
+        // Reset the state machine: store zero to unlock the first
+        // compare_exchange in the ISR, and consume any pending
+        // notification that we may have missed on a prior timeout.
+        self.isr_shared_state.timestamp.store(0, Ordering::Release);
+        let _ = self.interrupt_notification.wait(0);
+
+        // Drive the pin low to trigger the pulse, and ensure we put
+        // it back to high after our wait.
+        let trigger_pin = &mut self.trigger_pin;
+        trigger_pin.set_low()?;
+        defer! {
+            let _ = trigger_pin.set_high();
+        }
+
+        // Wait (up to timeout) for a notification from the
+        // ISR. Convert any result from the notification into a
+        // distance.
+        match self
+            .interrupt_notification
+            .wait(TickType::from(self.timeout).as_millis_u32())
```

Comment:
this call will suspend the executor thread meaning  no future will be able to advance during `self.timeout` how hard would it be to get a local copy of the executor and use a blocking call instead?

Reply:
~can we use a different executor entirely just to do this wait? What are the consequences if you just instantiated an instance of `smol::Executor`?~ Never mind, this was dumb.

Reply:
I brought this up in the micro-rdk team meeting a bit ago and we decided then that a blocking implementation was OK, at least initially. Of course we can change our minds about that and I'll see what I can do. However, I don't see an easy way to obtain access to the executor. Would you be OK merging with the blocking approach for now, and then filing a follow-up ticket to investigate how we can do it without blocking the executor? I know @gvaradarajan is also doing work in this space right now, so I'm interested to see what his investigation yields as well.

Reply:
Sorry I forgot in that case we should have a max for timeout_ms so someone cannot set a very large value accidentally and break things